nncf/torch/search_building_blocks/search_blocks.py

```diff
@@ -0,0 +1,561 @@
+from nncf.torch.nncf_network import NNCFNetwork
+
+
+IGNORED_NAME_OPERATORS = [*PTDropoutMetatype.get_all_aliases(), MODEL_OUTPUT_OP_NAME]
+
+class BuildingBlock:
+    def __init__(self, start_node, end_node):
+        self.start_node = start_node
+        self.end_node = end_node
+
+class BuildingBlockType(Enum): # only for BERT
+    MSHA = 'MSHA'
+    FF = 'FF'
+    Unknown = 'unknown'
+
+class BuildingBlockInfo:
+    def __init__(self, building_block: BuildingBlock,
+                       op_addresses: List[OperationAddress],
+                       modules: List[torch.nn.Module],
+                       block_type: BuildingBlockType):
+        self.building_block = building_block
+        self.op_addresses = op_addresses
+        self.modules = modules
+        self.block_type = block_type
+
+class SearchGraphNode:
+    """
+    Class describing nodes used in SGraph.
+    """
+
+    def __init__(self,
+                 node_key: str,
+                 data: dict = None):
+        self.node_key = node_key
+        self.data = data if data else {}
+
+    @property
+    def is_merged(self):
+        return self.data.get(SearchGraph.IS_MERGED_NODE_ATTR)
+
+    @property
+    def is_dummy(self):
+        return self.data.get(SearchGraph.IS_DUMMY_NODE_ATTR)
+
+    @property
+    def node_name(self) -> str:
+        return self.data.get(NNCFGraph.NODE_NAME_ATTR)
+
+    @property
+    def metatype(self) -> Type[OperatorMetatype]:
+        return self.data.get(NNCFGraph.METATYPE_ATTR)
+
+    @property
+    def node_type(self) -> str:
+        return self.data.get(NNCFGraph.NODE_TYPE_ATTR)
+
+    @property
+    def layer_name(self) -> str:
+        return self.data.get(NNCFGraph.LAYER_NAME_ATTR)
+
+    @property
+    def main_id(self) -> int:
+        # return the first id node
+        if not self.is_merged:
+            return self.data.get('id')
+        return self.data.get(SearchGraph.MERGED_NODES_NODE_ATTR)[0].get('id')
+
+    @property
+    def bottom_id(self) -> int:
+        if not self.is_merged:
+            return self.data.get('id')
+        return self.data.get(SearchGraph.MERGED_NODES_NODE_ATTR)[-1].get('id')
+
+    def __str__(self):
+        return self.node_key
+
+    def __hash__(self):
+        return hash(str(self))
+
+    def __eq__(self, other):
+        return isinstance(other, SearchGraphNode) and self.node_key == other.node_key
+
+
+class SearchGraph():
+    """
+    A wrapper over the graph, which represents the DNN execution graph transformed
+    by pattern matching, merging nodes and inserting auxiliary nodes.
+
+    """
+    ACTIVATION_INPUT_SHAPE_ATTR = 'activation_input_shape'
+    ACTIVATION_OUTPUT_SHAPE_ATTR = 'activation_output_shape'
+    IS_MERGED_NODE_ATTR = 'is_merged'
+    IS_DUMMY_NODE_ATTR = 'is_dummy'
+    KEY_NODE_ATTR = 'key'
+    MERGED_NODES_NODE_ATTR = 'merged_nodes'
+    TYPE_NODE_ATTR = 'type'
+    DUMMY_POSTFIX = " dummy"
+
+
+    def __init__(self, nx_merged_graph):
```

Comment:
Docstrings for every public method, type hints everywhere.

Reply:
Done.